lib/puppet/type.rb

```diff
@@ -2117,8 +2117,17 @@ def autorelation(rel_type, rel_catalog = nil)
     auto_rel = "eachauto#{rel_type}".to_sym
 
     self.class.send(auto_rel) { |type, block|
-      # Ignore any types we can't find, although that would be a bit odd.
-      next unless Puppet::Type.type(type)
+      if not Puppet::Type.type(type)
+        # We know it is not a type at this point, let's see if it exists as a defined type in the catalog
+        # We split and join on :: because defined types will likely have colon separators
+        # All defined types are of type :component
+        found_defines = rel_catalog.resources.select do |x|
+          x.type == :component && x.title_class == type.to_s.split('::').map(&:capitalize).join('::')
```

Comment:
@hlindberg it looks like we can get access to the `definition?` predicate method using:

```
(rdb:2) rel_catalog.resource('Ipset', 'foo').instance_variable_get(:@reference).resource_type.definition?
true
```

where ipset is the type of the defined type and foo is the title. However, we're having to use a private instance variable, and seems like there should be a better way. For example, Puppet::Type knows if it's an `application?`, but not if it's a `definition?`. Thoughts on how to handle that?